physical/cache.go

```diff
@@ -34,16 +43,58 @@ func NewCache(b Backend, size int, logger log.Logger) *Cache {
 	c := &Cache{
 		backend: b,
 		lru:     cache,
+		locks:   make(map[string]*sync.RWMutex, 256),
+		logger:  logger,
 	}
+	if err := locksutil.CreateLocks(c.locks, 256); err != nil {
+		logger.Error("physical/cache: error creating locks", "error", err)
+		return nil
+	}
+
+	if txnl, ok := c.backend.(Transactional); ok {
+		c.transactional = txnl
+	}
+
 	return c
 }
 
+func (c *Cache) lockHashForKey(key string) string {
+	hf := sha1.New()
```

Comment:
Keys can be any arbitrary string, so hashing and taking the first byte ensures that we can make a map of keys that will always work for any key value. Similar to what the token store does, except there's no "global" fallback lock since basically all keys would otherwise hit it.